gusto/timeloop.py

```diff
@@ -138,108 +172,245 @@ def run(self, t, tmax, pickup=False):
+
+        super().__init__(state,
+                         equations_schemes,
+                         physics_list=physics_list,
+                         prescribed_fields=prescribed_fields)
+
+    @property
+    def advecting_velocity(self):
+        try:
+            return self.state.fields("u")
+        except AttributeError:
+            raise ValueError("You have not specified an advecting velocity to use")
+
+
+class CrankNicolson(Timestepper):
     """
     This class implements a Crank-Nicolson discretisation, with Strang
     splitting and auxilliary semi-Lagrangian advection.
+    Defines the advecting velocity to be the average of u_n and u_{n+1}.
+    After applying the semi implicit step passively advected fields are
+    advected and any diffusion terms are applied.
 
     :arg state: a :class:`.State` object
+    :arg equation_set: (optional) a :class:`.PrognosticMixedEquation` object,
+    the main set of equations to be solved (eg ShallowWaterEquations)
     :arg advected_fields: iterable of ``(field_name, scheme)`` pairs
         indicating the fields to advect, and the
         :class:`~.Advection` to use.
-    :arg linear_solver: a :class:`.TimesteppingSolver` object
-    :arg forcing: a :class:`.Forcing` object
-    :arg diffused_fields: optional iterable of ``(field_name, scheme)``
-        pairs indictaing the fields to diffusion, and the
+    :arg diffused_fields: (optional) iterable of ``(field_name, scheme)``
+        pairs indictaing the fields to diffuse, and the
         :class:`~.Diffusion` to use.
+    :arg equations_schemes: a list of tuples (equation, schemes)
+         pairing a :class:`.PrognosticEquation` object with the scheme(s)
+         to use to timestep it. The schemes entry can be a :class:`.Advection`
+         object which is applied to the entire equation, or a tuple of
+         (scheme, *active_labels) where scheme is a :class:`.Advection`
+         object and the active_labels are :class:`.Label` objects that
+         specify which parts of the equation to apply this scheme to.
     :arg physics_list: optional list of classes that implement `physics` schemes
     :arg prescribed_fields: an order list of tuples, pairing a field name with a
          function that returns the field as a function of time.
+    :kwargs: maxk is the number of outer iterations, maxi is the number of inner
+             iterations and alpha is the offcentering parameter
     """
 
-    def __init__(self, state, advected_fields, linear_solver, forcing,
-                 diffused_fields=None, physics_list=None, prescribed_fields=None):
+    def __init__(self, state, equation_set, advected_fields,
+                 diffused_fields=None, equations_schemes=None,
+                 physics_list=None, prescribed_fields=None, **kwargs):
+
+        self.maxk = kwargs.pop("maxk", 4)
+        self.maxi = kwargs.pop("maxi", 1)
+        self.alpha = kwargs.pop("alpha", 0.5)
```

Comment:
I think we should start trying to `Constant`ize coefficients that we know won't change throughout a run. That way, if we want to change the parameters and rerun a simulation, we won't have to regenerate and recompile code.

Reply:
I think I was pretty careful to wrap `Constant` around things when the forms are manipulated. In fact, if you look in the `__rmul__` method of `Term` you'll see that I wrap floats with `Constant` so even if someone forgets the `Constant` it'll be done anyway.